### PR TITLE
Update: process monitor events all at once upon commit

### PR DESCRIPTION
**Pull request by Ignatella**
Hi!

While using nftables, we discovered it would be more efficient to process monitor events (changes) all at once upon commit, triggered by the event that caused the change, such as [NFT_MSG_NEWGEN](https://elixir.bootlin.com/linux/v6.10.10/source/net/netfilter/nf_tables_api.c#L10514).

This PR modifies the way events are handled by grouping all changes together and processing them in response to the event that initiated the change. Additionally, the event header, which includes important information such as the PID, is now propagated. This allows us to identify the process responsible for the change.

### CODE REVIEW

File: set.go [hunk 1]
```diff
@@ -797,10 +803,16 @@ func elementsFromMsg(fam byte, msg netlink.Message) ([]SetElement, error) {
 	}
 	ad.ByteOrder = binary.BigEndian
 
+	var info = &SetElementsInfo{}
```
Comment (patryk4815) [inline set.go:806]: missing return?

File: monitor.go [hunk 2]
```diff
@@ -293,7 +331,7 @@ func (monitor *Monitor) Close() error {
 // calling Close on Monitor or encountering a netlink conn error while Receive.
 // Caller may receive a MonitorEventTypeOOB event which contains an error we didn't
 // handle, for now.
-func (cc *Conn) AddMonitor(monitor *Monitor) (chan *MonitorEvent, error) {
+func (cc *Conn) AddMonitor(monitor *Monitor) (chan *MonitorEvents, error) {
```
Comment (stapelberg) [inline monitor.go:334]: There are existing users of this method, so we should not change its API.

Maybe we could introduce a new AddGenerationalMonitor() method with your suggested signature, and make AddMonitor work with generations, unpack the events, put them on a channel of the old type?

File: monitor.go [hunk 3]
```diff
@@ -192,20 +198,30 @@ func NewMonitor(opts ...MonitorOption) *Monitor {
 }
 
 func (monitor *Monitor) monitor() {
+	changesEvents := make([]*MonitorEvent, 0, 2)
```
Comment (stapelberg) [inline monitor.go:201]: Here and below: Why initialize the slice with capacity 2? It seems unlikely to me that this would make a difference. The slice’s capacity will quickly grow to the maximum needed and then stay there, I think.

File: monitor.go [hunk 4]
```diff
@@ -116,9 +116,15 @@ const (
...
-	Data  any
-	Error error
+	Header netlink.Header
+	Type   MonitorEventType
+	Data   any
+	Error  error
+}
+
+type MonitorEvents struct {
+	GenerateBy *MonitorEvent
```
Comment (stapelberg) [inline monitor.go:126]: I think `GeneratedBy` (with a d) reads better.

File: set.go [hunk 5]
```diff
@@ -268,6 +268,12 @@ type Set struct {
 	KeyByteOrder binaryutil.ByteOrder
 }
 
+type SetElementsInfo struct {
```
Comment (stapelberg) [inline set.go:271]: What’s the relation between this part of the change and the rest of it? Could the set changes go into a separate commit / PR?

**Review by stapelberg — CHANGES_REQUESTED** [review]

**Comment by stapelberg** [comment]
cc @singchia who contributed the monitoring code

File: set.go (hunk 1 above)
Comment (Ignatella) [reply set.go]: yes, thank You

File: monitor.go (hunk 2 above)
Comment (Ignatella) [reply monitor.go]: AddGenerationalMonitor introduced, thank You

File: monitor.go (hunk 3 above)
Comment (Ignatella) [reply monitor.go]: switched to

```go
make([]*MonitorEvent, 0)
```

File: monitor.go (hunk 4 above)
Comment (Ignatella) [reply monitor.go]: switched to `GeneratedBy`

File: set.go (hunk 5 above)
Comment (Ignatella) [reply set.go]: Removed from the PR, I will propose this change as part of another PR. Thank You

File: monitor.go [hunk 6]
```diff
@@ -192,20 +198,30 @@ func NewMonitor(opts ...MonitorOption) *Monitor {
 }
 
 func (monitor *Monitor) monitor() {
+	changesEvents := make([]*MonitorEvent, 0)
```
Comment (stapelberg) [inline monitor.go:201]: Write `var changesEvents []*MonitorEvent` here. `make(…, 0)` is equivalent to just declaring a variable.

File: monitor.go [hunk 7]
```diff
@@ -192,20 +198,30 @@ func NewMonitor(opts ...MonitorOption) *Monitor {
...
 				}
-				monitor.eventCh <- event
+
+				changesEvents = append(changesEvents, event)
+
+				monitor.eventCh <- &MonitorEvents{
+					GeneratedBy: event,
+					Changes:     changesEvents,
+				}
+				changesEvents = make([]*MonitorEvent, 0)
```
Comment (stapelberg) [inline monitor.go:223]: This can just be `changesEvents = nil`

File: monitor.go [hunk 8]
```diff
@@ -294,6 +332,27 @@ func (monitor *Monitor) Close() error {
 // Caller may receive a MonitorEventTypeOOB event which contains an error we didn't
 // handle, for now.
 func (cc *Conn) AddMonitor(monitor *Monitor) (chan *MonitorEvent, error) {
+	generationalEventCh, err := cc.AddGenerationalMonitor(monitor)
+
```
Comment (stapelberg) [inline monitor.go:336]: nit: remove blank line between err assignment and err check

**Review by stapelberg — CHANGES_REQUESTED** [review]

File: monitor.go (hunk 6 above)
Comment (Ignatella) [reply monitor.go]: changed to declaration. Thank you

File: monitor.go (hunk 7 above)
Comment (Ignatella) [reply monitor.go]: Changed to `changesEvents = nil`

File: monitor.go (hunk 8 above)
Comment (Ignatella) [reply monitor.go]: Removed

**Comment by Ignatella** [comment]
Also added ProcComm check to `TestMonitor`

File: monitor_test.go [hunk 9]
```diff
@@ -66,23 +69,35 @@ func TestMonitor(t *testing.T) {
 			if !ok {
 				return
 			}
-			if event.Error != nil {
-				err = fmt.Errorf("monitor err: %s", event.Error)
```
Comment (patryk4815) [inline monitor_test.go:70]: the original version of the test seems to be broken, "err" is not checked at all

File: monitor_test.go [hunk 10]
```diff
@@ -66,23 +69,35 @@ func TestMonitor(t *testing.T) {
...
+					gotTable = change.Data.(*nftables.Table)
+					atomic.AddInt32(&count, 1)
+				case nftables.MonitorEventTypeNewChain:
+					gotChain = change.Data.(*nftables.Chain)
+					atomic.AddInt32(&count, 1)
+				case nftables.MonitorEventTypeNewRule:
+					gotRule = change.Data.(*nftables.Rule)
+					atomic.AddInt32(&count, 1)
+				}
+				if atomic.LoadInt32(&count) == 3 {
```
Comment (patryk4815) [inline monitor_test.go:98]: this check looks bad to me, should throw error when count != 3

File: monitor.go [hunk 11]
```diff
@@ -221,54 +237,76 @@ func (monitor *Monitor) monitor() {
...
+					Header: msg.Header,
 				}
-				monitor.eventCh <- event
+
+				monitor.eventCh <- &MonitorEvents{
+					GeneratedBy: event,
+					Changes:     changesEvents,
+				}
+
+				changesEvents = make([]*MonitorEvent, 0)
```
Comment (stapelberg) [inline monitor.go:305]: This should also be `changesEvents = nil`

**Review by stapelberg — COMMENTED** [review]
Almost ready to merge

File: monitor.go (hunk 11 above)
Comment (Ignatella) [reply monitor.go]: Changed, thank You

**Review by stapelberg — APPROVED** [review]
